python/lsst/daf/butler/remote_butler/server/handlers/_external_query.py

```diff
@@ -29,40 +29,85 @@
-from contextlib import contextmanager
+from collections.abc import AsyncIterator, Iterable, Iterator
+from contextlib import ExitStack, contextmanager
 from typing import NamedTuple
 
 from fastapi import APIRouter, Depends
+from fastapi.concurrency import contextmanager_in_threadpool, iterate_in_threadpool
+from fastapi.responses import StreamingResponse
 from lsst.daf.butler import DataCoordinate, DimensionGroup
 from lsst.daf.butler.remote_butler.server_models import (
     QueryAnyRequestModel,
     QueryAnyResponseModel,
     QueryCountRequestModel,
     QueryCountResponseModel,
     QueryExecuteRequestModel,
-    QueryExecuteResponseModel,
     QueryExplainRequestModel,
     QueryExplainResponseModel,
     QueryInputs,
 )
 
-from ....queries.driver import QueryDriver, QueryTree
+from ....queries.driver import QueryDriver, QueryTree, ResultPage, ResultSpec
 from .._dependencies import factory_dependency
 from .._factory import Factory
-from ._query_serialization import convert_query_pages
+from ._query_serialization import serialize_query_pages
 
 query_router = APIRouter()
 
 
 @query_router.post("/v1/query/execute", summary="Query the Butler database and return full results")
 def query_execute(
     request: QueryExecuteRequestModel, factory: Factory = Depends(factory_dependency)
-) -> QueryExecuteResponseModel:
-    with _get_query_context(factory, request.query) as ctx:
+) -> StreamingResponse:
+    # Managing the lifetime of the query context object is a little tricky.  We
+    # need to enter the context here, so that we can immediately deal with any
+    # exceptions raised by query set-up.  We eventually transfer control to an
+    # iterator consumed by FastAPI's StreamingResponse handler, which will
+    # start iterating after this function returns.  So we use this ExitStack
+    # instance to hand over the context manager to the iterator.
+    with ExitStack() as exit_stack:
+        ctx = exit_stack.enter_context(_get_query_context(factory, request.query))
         spec = request.result_spec.to_result_spec(ctx.driver.universe)
-        pages = ctx.driver.execute(spec, ctx.tree)
-        return QueryExecuteResponseModel(result=convert_query_pages(spec, pages))
+        response_pages = ctx.driver.execute(spec, ctx.tree)
+
+        # We write the response incrementally, one page at a time, as
+        # newline-separated chunks of JSON.  This allows clients to start
+        # reading results earlier and prevents the server from exhausting
+        # all its memory buffering rows from large queries.
+        output_generator = _stream_query_pages(
+            # Transfer control of the context manager to
+            # _stream_query_pages.
+            exit_stack.pop_all(),
+            spec,
+            response_pages,
+        )
+        return StreamingResponse(output_generator, media_type="application/jsonlines")
+
+    # Mypy thinks that ExitStack might swallow an exception.
+    assert False, "This line is unreachable."
```

Comment:
I presume there is no other more explicit way to suppress this with `# type: ignore[something]`?

Reply:
Not that I'm aware of, the issue is pretty basic to how mypy analyzes control flow.

Most context managers are declared with a return type of `Literal[False]` so that mypy knows they won't swallow an exception, but because ExitStack could contain any context manager it returns a `bool` and mypy has to assume that we might bypass the return statement inside the context manager.